spanner/batchwriter.go

```diff
@@ -0,0 +1,320 @@
+	for k, v := range bw.a.errors {
+		m[k] = v
+	}
+	return m
+}
+
+func (bw *BatchWriter) getBadRowsForTest() []*row {
+	return bw.a.badRows
+}
+
+// getBatch returns a slice of data from the front of bw.rows.  The slice
+// returned is the largest one not exceeding countThreshold and byteThreshold.
+func (bw *BatchWriter) getBatch() (rows []*row, count int64, bytes int64) {
+	for i, _ := range bw.rows {
+		c := count + int64(len(bw.rows[i].cols))
+		b := bytes + byteSize(bw.rows[i])
+		// If next row puts us over the thresholds, then stop. But make sure
+		// we have at least one row. If a single row puts us over the
+		// thresholds, there's not much we can do: we just try sending it to Spanner
+		// (it might succeed, since our thresholds are conservative).
+		if (c >= countThreshold || b >= byteThreshold) && len(rows) >= 1 {
+			bw.rCount -= count
+			bw.rBytes -= bytes
+			bw.rows = bw.rows[i:]
+			return rows, count, bytes
+		}
+		count = c
+		bytes = b
+		rows = append(rows, bw.rows[i])
+	}
+	bw.rCount = 0
+	bw.rBytes = 0
+	bw.rows = nil
+	return rows, count, bytes
+}
+
+func (bw *BatchWriter) errorStats(rows []*row, err error, retry bool) {
+	if bw.verbose {
+		fmt.Printf("Error while writing %d rows to Spanner: %v\n", len(rows), err)
+	}
+
+	bw.a.lock.Lock()
+	defer bw.a.lock.Unlock()
+
+	bw.a.errors[err.Error()]++
+	if retry {
+		return
+	}
+	// All rows in r will be dropped.
+	if len(rows) == 1 {
+		// This is a confirmed bad row: add it to the badRows list.
+		r := rows[0]
+		n := byteSize(r)
+		// Use bw.bytesLimit to cap storage used by badRows. Keep at least one bad row.
+		if bw.a.badRowsBytes+n < bw.bytesLimit || len(bw.a.badRows) == 0 {
+			bw.a.badRows = append(bw.a.badRows, r)
+			bw.a.badRowsBytes += n
+		}
+	}
+	for _, x := range rows {
+		bw.a.droppedRows[x.table]++
+	}
+	return
+}
+
+// Note: doWriteAndHandleErrors must be thread-safe.
+func (bw *BatchWriter) doWriteAndHandleErrors(rows []*row) {
+	var m []*sp.Mutation
+	for _, x := range rows {
+		m = append(m, sp.InsertOrUpdate(x.table, x.cols, x.vals))
+	}
+	if err := bw.write(m); err != nil {
+		hitRetryLimit := atomic.LoadInt64(&bw.a.retries) >= bw.retryLimit
+		retry := len(rows) > 1 && !hitRetryLimit
+		bw.errorStats(rows, err, retry)
+		if !retry {
+			if hitRetryLimit && bw.verbose {
+				fmt.Printf("Have hit %d retries: will no do any more\n", atomic.LoadInt64(&bw.a.retries))
+			}
+			return
+		}
+		// Split into 10 pieces and retry. This is useful
+		// if a batch contains a bad data row (Spanner
+		// will fail the entire batch). In effect we attempt
+		// to narrow down which row (or rows) are bad, and
+		// write the 'good' rows to Spanner.
+		k := 1 + len(rows)/10
+		min := func(i, j int) int {
+			if i <= j {
+				return i
+			}
+			return j
+		}
+		for i := 0; i < len(rows); i += k {
+			atomic.AddInt64(&bw.a.retries, 1)
+			bw.doWriteAndHandleErrors(rows[i:min(i+k, len(rows))])
+		}
+	}
+}
+
+// Note: backgroundWrite must be thread-safe.
```

Comment:
"must be thread-safe" -- why?

Reply:
backgroundWrite and doWriteAndHandleErrors are run via go-routines.
So there will be multiple instances of these functions running.

The boundary between single-threaded vs go routines is the call to go bw.backgroundWrite in startWrite.

Reply:
Might be worth mentioning that in the comment?

Reply:
Done. Comment for this function now reads:
// Note: backgroundWrite must be thread-safe because it is run as
// a go routine.

Added similar comment to previous function.